File: java/examples/cc0mon-api-get-image-svg.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17
//DEPS com.cc0mon:sdk:0.1.0

import com.cc0mon.sdk.Client;
import com.cc0mon.sdk.Errors;
import com.cc0mon.sdk.LoggingSetup;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.logging.Logger;

class GetImageSvg {
    public static void main(String[] args) throws Exception {
        Logger log = LoggingSetup.configure("get-image-svg");
        Integer id = null;
        Path out = null;
        for (int i = 0; i + 1 < args.length; i++) {
            if ("--id".equals(args[i])) try { id = Integer.parseInt(args[i + 1]); } catch (NumberFormatException e) {}
            if ("--out".equals(args[i])) out = Paths.get(args[i + 1]);
        }
        if (id == null) { System.err.println("usage: cc0mon-api-get-image-svg.java --id <1..10000> [--out path]"); System.exit(5); }
        if (out == null) out = Paths.get(System.getProperty("user.dir"), "cc0mon-" + id + ".svg");
        log.info("script start id=" + id + " out=" + out);

        try (Client client = new Client()) {
            byte[] bytes = client.getImageSvg(id);
            Files.write(out, bytes);
            System.out.println(out.toAbsolutePath());
            log.info("script exit code=0 bytes=" + bytes.length + " path=" + out);
        } catch (Errors.ValidationException e) { log.severe("validation: " + e.getMessage()); System.err.println("validation error: " + e.getMessage()); System.exit(5);
        } catch (Errors.NetworkException e)    { log.severe("network: " + e.getMessage());    System.err.println("network error: " + e.getMessage()); System.exit(2);
        } catch (Errors.RateLimitException e)  { log.severe("rate limited http " + e.getStatusCode()); System.err.println("rate limited HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(3);
        } catch (Errors.ClientApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("client error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(3);
        } catch (Errors.ServerApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("server error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(4);
        } catch (Exception e) { log.severe("unexpected: " + e); System.err.println("unexpected: " + e); System.exit(1); }
    }
}

File: python/examples/cc0mon-api-get-image-svg.py
"""cc0mon-api-get-image-svg — download the SVG artwork for a token.

Usage::

    python cc0mon-api-get-image-svg.py --id 1
    python cc0mon-api-get-image-svg.py --id 1 --out .\\my-mon.svg

Output is silently overwritten if it exists.
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

from cc0mon_sdk import (
    Client,
    ClientApiError,
    NetworkError,
    ServerApiError,
    ValidationError,
    configure_logger,
)

ACTION = "get-image-svg"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Download a cc0mon SVG image.")
    parser.add_argument("--id", dest="token_id", type=int, required=True, help="Token id (1-10000)")
    parser.add_argument("--out", dest="out_path", type=Path, default=None, help="Output path (default ./cc0mon-<id>.svg)")
    args = parser.parse_args(argv)

    out_path = args.out_path or Path.cwd() / f"cc0mon-{args.token_id}.svg"

    logger = configure_logger(ACTION)
    logger.info("script start id=%d out=%s", args.token_id, out_path)

    try:
        with Client() as client:
            svg_bytes = client.get_image_svg(args.token_id)
    except ValidationError as e:
        logger.error("validation: %s", e)
        print(f"validation error: {e}", file=sys.stderr)
        return 5
    except NetworkError as e:
        logger.error("network: %s", e)
        print(f"network error: {e}", file=sys.stderr)
        return 2
    except ClientApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"client error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 3
    except ServerApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"server error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 4

    out_path.write_bytes(svg_bytes)
    print(out_path.resolve())
    logger.info("script exit code=0 bytes=%d path=%s", len(svg_bytes), out_path)
    return 0


if __name__ == "__main__":
    sys.exit(main())
